#167729788 setup social media (Google and Facebook) login

sebaztyn opened the pull request:
#### What does this PR do?
Set up social media (Google and Facebook) logins on the account
#### Description of Task to be completed?
- install passport module
- install passport-google-oauth2 module
- install passport-facebook module
- set up the passport processes to enable for seamless integration of social media processes in logging in

#### How should this be manually tested?
 - clone repository
- startup terminal
- cd into the cloned repository
- start the server by typing ```npm run dev``` into the terminal
- visit a browser and type ```http://localhost:3000/api/v1/auth/google/login``` or
- visit a browser and type ```http://localhost:3000/api/v1/auth/facebook/login```

#### Any background context you want to provide?
None

#### What are the relevant pivotal tracker stories?
[#167729788](https://www.pivotaltracker.com/story/show/167729788)

#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/36090168/63550138-f9653000-c529-11e9-879c-7cfc98cd2397.png)

#### Questions:
None

sebaztyn committed:
feature(Server): setup social media login

- setup google login for the project
- setup facebook login for the project

[Finishes #167729788]

okabamac reviewed (commented):
Great work sir, just try to fix the travis and the hounds let's see

sebaztyn commented:
@okabamac Thank you for your feedback. The travis will likely not clear until this pull request is closed as this functionality is tied to the heroku staging domain.
Secondly, one of the 4 hound errors will also remain as I am only importing and not using the imported module but on importing it to allow for its continuous use within the particular work file

chibytez reviewed (approved):
Nice work

Stylll commented on .eslintrc.json:
```diff
@@ -46,6 +48,10 @@
       }
     ]
   },
+    "plugins": [
```
is there a reason we are adding frontend plugin to this backend app?

sebaztyn replied:
I have taken it out

Stylll commented on package.json:
```diff
@@ -11,14 +11,15 @@
     "codeclimate-coverage": "codeclimate-test-reporter < lcov.info",
     "generate-lcov": "nyc report --reporter=text-lcov > lcov.info",
-    "test": "cross-env NODE_ENV=test mocha --require @babel/register src/test/**.test.js --timeout 20000 --exit"
+    "test": "cross-env NODE_ENV=test npm run heroku-postbuild && nyc _mocha --require @babel/register src/test/**.test.js --timeout 20000 --exit"
```
Hey @sebaztyn what's the difference between `mocha` and `_mocha`?. Also, why are we running `heroku-postbuild` in the test script?.

sebaztyn replied:
Thank you for reviewing my work. ```nyc _mocha``` was something I learnt in an old documentation. It also works just fine with ```nyc mocha```; I have taken out the ```_```.
our script ```heroku-postbuild``` runs a combined script that creates, migrates and seeds data into the table which was why I used it.

Stylll commented on src/database/migrations/20190818182539-create-user.js:
```diff
@@ -4,9 +4,13 @@ export default {
+      type: Sequelize.STRING,
+      allowNull: false,
+      primaryKey: true
+    },
```
Hey @sebaztyn, the changes you made in this file has to be in a new migration file. The point of having migration files is to have a trackable process of updates done to the database so as to easily rollback changes when needed. Also, making a change in this file means when this feature gets to staging, prod or any other local environment asides yours, the DB and migration tracker will have to be cleared so as to make this migration run again.

sebaztyn replied:
@Stylll thank you for the feedback. I have effected the changes as advised.
Thank you.

Stylll reviewed (changes requested):
Hey @sebaztyn, I left a few comments. Please check them out.

sebaztyn committed:
feature(endpoint): Implement Product Owner's Feedback

- socialmedia login should be implemented for only staff
- return appropriate messages for the different use cases

[Finishes #167729788]

Stylll commented on package.json:
```diff
@@ -10,15 +10,15 @@
     "codeclimate-coverage": "codeclimate-test-reporter < lcov.info",
     "generate-lcov": "nyc report --reporter=text-lcov > lcov.info",
-    "test": "cross-env NODE_ENV=test mocha --require @babel/register src/test/**.test.js --timeout 20000 --exit"
+    "test": "cross-env NODE_ENV=test npm run heroku-postbuild && nyc mocha --require @babel/register src/test/**.test.js --timeout 20000 --exit"
```
@sebaztyn  That script is strictly for heroku-postbuild, we shouldn't use it here so we don't call it unnecessary if it is later updated with something specifically for Heroku. You can call the scripts to undo migration, run migration and seed test here or create a new script combining them together.

Check out @chibytez 's PR. He has a script for that

sebaztyn replied:
Thank you @Stylll . I have now implemented the recommended pattern

okafor-chidimma reviewed (approved):
Good work..

Stylll commented on src/routes/api/index.js:
```diff
@@ -1,8 +1,9 @@
-
+usersRouter.use('/', userRoutes);
 usersRouter.post('/users/login', (req, res, next) => {
   if (!req.body.user.email) {
```
@sebaztyn We should try to avoid write middlewares or controller code inside the route. This validation should be done in a middleware and used here

sebaztyn replied:
@Stylll , thank you for the feedback. I have taken off the code. That was a code that came with the cloned codebase.

Stylll commented on .env.example:
```diff
@@ -1,2 +1,15 @@
+CLIENTID_PROD = ''
+CLIENTSECRET_PROD = ''
+GOOGLE_CALLBACK_URL_PROD = ''
+CLIENTID_TEST = ''
```
@sebaztyn we should the remove the spaces in-between so we don't confuse the users. `CLIENTID_TEST = ''` should be like `CLIENTID_TEST=''`

sebaztyn replied:
I made made the corrections are advised

Stylll commented on src/routes/api/index.js:
```diff
@@ -1,8 +1,9 @@
+usersRouter.use('/', userRoutes);
 usersRouter.post('/users/login', (req, res, next) => {
   if (!req.body.user.email) {
     return res.status(422).json({ errors: { email: "can't be blank" } });
```
@sebaztyn Let's stick to the status code of `400` here. I know there's a lot of debate on how and where to use `422` instead but we should stick to `400`. Also, the error message should be more like `email can't be blank` and not `can't be blank` so users can easily understand what entity the message is referring

sebaztyn replied:
Dear @Stylll, I have made changes. Thank you.

Stylll reviewed (changes requested):
@sebaztyn I left a few comments. Please check them out. 
I also noticed there's no test for this feature. Please create a test suite.

Please update the label to `Peer Review`

sebaztyn committed:
Merge branch 'staging' of https://github.com/andela/shadowcat-backend into feature-social-media-login-167729788

okafor-chidimma reviewed (approved):
All good from here...well done
